Get column value by column label in Framework Core ASoA

mytkom opened the pull request:
I will continue our discussion here @ktf, @jgrosseo, @saganatt.

@jgrosseo was correct; this solution shouldn't be implemented as it is. It was slower in this benchmark by 3-10 times for simple examples. I will leave this draft PR here and try to implement this alternative solution I was writing about in https://github.com/AliceO2Group/O2Physics/pull/7371.  That 20% I was speaking of during the presentation was the whole task performance difference locally on my laptop. These are my local benchmark results:
```bash
------------------------------------------------------------------------------------------------------------
Benchmark                                                  Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------
BM_ASoADynamicColumnPresent/8                            237 ns          236 ns      2983783 bytes_per_second=258.465M/s
BM_ASoADynamicColumnPresent/64                           395 ns          395 ns      1717027 bytes_per_second=1.20857G/s
BM_ASoADynamicColumnPresent/512                         1559 ns         1555 ns       448023 bytes_per_second=2.45328G/s
BM_ASoADynamicColumnPresent/4096                       10961 ns        10933 ns        62651 bytes_per_second=2.79137G/s
BM_ASoADynamicColumnPresent/32768                      85382 ns        85164 ns         8024 bytes_per_second=2.86673G/s
BM_ASoADynamicColumnPresent/262144                    679514 ns       677289 ns          998 bytes_per_second=2.88374G/s
BM_ASoADynamicColumnPresentGetValueByLabel/8             419 ns          418 ns      1622625 bytes_per_second=145.876M/s
BM_ASoADynamicColumnPresentGetValueByLabel/64           1772 ns         1768 ns       392634 bytes_per_second=276.255M/s
BM_ASoADynamicColumnPresentGetValueByLabel/512         12671 ns        12641 ns        54994 bytes_per_second=309.019M/s
BM_ASoADynamicColumnPresentGetValueByLabel/4096        99301 ns        99056 ns         6893 bytes_per_second=315.478M/s
BM_ASoADynamicColumnPresentGetValueByLabel/32768      805034 ns       802857 ns          858 bytes_per_second=311.388M/s
BM_ASoADynamicColumnPresentGetValueByLabel/262144    6319410 ns      6301477 ns          107 bytes_per_second=317.386M/s
BM_ASoADynamicColumnCall/8                              25.3 ns         25.2 ns     28377608 bytes_per_second=2.36521G/s
BM_ASoADynamicColumnCall/64                              183 ns          182 ns      3722494 bytes_per_second=2.61903G/s
BM_ASoADynamicColumnCall/512                            1378 ns         1375 ns       478496 bytes_per_second=2.7751G/s
BM_ASoADynamicColumnCall/4096                          11039 ns        11012 ns        62893 bytes_per_second=2.77126G/s
BM_ASoADynamicColumnCall/32768                         88298 ns        88067 ns         7847 bytes_per_second=2.77221G/s
BM_ASoADynamicColumnCall/262144                       706708 ns       704322 ns          996 bytes_pelementaryer_second=2.77306G/s
BM_ASoADynamicColumnCallGetValueByLabel/8               67.6 ns         67.4 ns     10002511 bytes_per_second=905.098M/s
BM_ASoADynamicColumnCallGetValueByLabel/64               500 ns          498 ns      1345915 bytes_per_second=979.614M/s
BM_ASoADynamicColumnCallGetValueByLabel/512             3981 ns         3972 ns       175994 bytes_per_second=983.554M/s
BM_ASoADynamicColumnCallGetValueByLabel/4096           31831 ns        31757 ns        22014 bytes_per_second=984.025M/s
BM_ASoADynamicColumnCallGetValueByLabel/32768         255368 ns       254706 ns         2722 bytes_per_second=981.526M/s
BM_ASoADynamicColumnCallGetValueByLabel/262144       2029059 ns      2022533 ns          340 bytes_per_second=988.859M/s
```

mytkom committed:
use generic get() in row_helpers

mytkom committed:
backward compatibility for binning policy

mytkom committed:
getColumnValueByLabel function

mytkom committed:
reverse logic

mytkom committed:
benchmark for getColumnValueByLabel

mytkom committed:
formatting adjustments

mytkom committed:
getColumnGetterByLabel and benchmarks

mytkom committed:
formatting adjustments

ktf commented:
Thank you for your investigation. Much appreciated.

As a general note, please try to have a smaller contributions if possible, because that will for sure simplify and speedup the review process.

mytkom committed:
improve performance by not using std::function

alibuild committed:
Please consider the following formatting changes

mytkom commented:
Previously, I used `std::function` as the return type for the getter, but the performance was pretty bad. I changed O2 to debug mode and used perf + speedscope to analyze it. I obtained a much better performance, almost equal to an explicit getter call (less than 10% execution time difference). Then I switched back to release mode for compilation, and these are the results:
```
-------------------------------------------------------------------------------------------------------------
Benchmark                                                   Time             CPU   Iterations UserCounters...
-------------------------------------------------------------------------------------------------------------
BM_ASoADynamicColumnPresent/8                             236 ns          235 ns      2951784 bytes_per_second=259.24M/s
BM_ASoADynamicColumnPresent/64                            397 ns          396 ns      1765157 bytes_per_second=1.2049G/s
BM_ASoADynamicColumnPresent/512                          1557 ns         1553 ns       454082 bytes_per_second=2.45591G/s
BM_ASoADynamicColumnPresent/4096                        10783 ns        10758 ns        63512 bytes_per_second=2.83666G/s
BM_ASoADynamicColumnPresent/32768                       85087 ns        84863 ns         8263 bytes_per_second=2.87686G/s
BM_ASoADynamicColumnPresent/262144                     683696 ns       681478 ns         1030 bytes_per_second=2.86601G/s
BM_ASoADynamicColumnPresentGetGetterByLabel/8             277 ns          276 ns      2529198 bytes_per_second=221.199M/s
BM_ASoADynamicColumnPresentGetGetterByLabel/64            658 ns          656 ns      1077150 bytes_per_second=744.385M/s
BM_ASoADynamicColumnPresentGetGetterByLabel/512          3490 ns         3482 ns       199110 bytes_per_second=1121.89M/s
BM_ASoADynamicColumnPresentGetGetterByLabel/4096        27125 ns        27054 ns        26808 bytes_per_second=1.12801G/s
BM_ASoADynamicColumnPresentGetGetterByLabel/32768      208795 ns       208255 ns         3366 bytes_per_second=1.17232G/s
BM_ASoADynamicColumnPresentGetGetterByLabel/262144    1668853 ns      1663740 ns          423 bytes_per_second=1.17394G/s
BM_ASoADynamicColumnCall/8                               24.8 ns         24.7 ns     28070062 bytes_per_second=2.41155G/s
BM_ASoADynamicColumnCall/64                               180 ns          180 ns      3630937 bytes_per_second=2.65376G/s
BM_ASoADynamicColumnCall/512                             1377 ns         1374 ns       497025 bytes_per_second=2.77621G/s
BM_ASoADynamicColumnCall/4096                           10961 ns        10935 ns        63683 bytes_per_second=2.79084G/s
BM_ASoADynamicColumnCall/32768                          87853 ns        87642 ns         8035 bytes_per_second=2.78565G/s
BM_ASoADynamicColumnCall/262144                        703888 ns       701646 ns          982 bytes_per_second=2.78363G/s
BM_ASoADynamicColumnCallGetGetterByLabel/8               40.0 ns         39.9 ns     17621092 bytes_per_second=1.4949G/s
BM_ASoADynamicColumnCallGetGetterByLabel/64               299 ns          298 ns      2335159 bytes_per_second=1.60043G/s
BM_ASoADynamicColumnCallGetGetterByLabel/512             2400 ns         2394 ns       300728 bytes_per_second=1.59373G/s
BM_ASoADynamicColumnCallGetGetterByLabel/4096           18681 ns        18635 ns        36845 bytes_per_second=1.63765G/s
BM_ASoADynamicColumnCallGetGetterByLabel/32768         148504 ns       148136 ns         4728 bytes_per_second=1.64809G/s
BM_ASoADynamicColumnCallGetGetterByLabel/262144       1189493 ns      1185832 ns          571 bytes_per_second=1.64705G/s
```
I am currently stuck with this solution; I would like to know if using the get() explicitly column struct member method is possible. I was trying to implement such an approach, but there is a type problem; the member method pointer type is ex. `float(*ColumnType)(const RowItType&)`, and I would know ColumnType only in runtime after string comparison, so I can't think of an efficient implementation.

I would like to know if you think such a performance difference would be acceptable and if you can think of a more efficient solution.

mytkom committed:
remove event mixing changes from this branch

mytkom commented:
According to @ktf's comment, I now split this PR into 2; this one would only implement getting column values by string, and event mixing adjustments are not added anymore.

mytkom committed:
add forgotten functions

mytkom committed:
fix column name condition for dynamic columns

mytkom commented:
@ktf @jgrosseo,
May you have a look at the latest changes?

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1988,63 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, typename... Cs>
+ColumnGetterFunction<R, T> getColumnGetterByLabel(o2::framework::pack<Cs...>, const std::string_view& columnLabel)
+{
+  std::string_view labelView(columnLabel);
```
Why do you actually need this?

mytkom replied:
It would be more flexible to choose columns based on JSON file prepared for every trained neural network in CCDB, than hardcode columns and need to handle many networks to be retrained after future neural network design change. It is also possible to use this function to prepare a more common interface for NN inference for other ML projects in ALICE than ours.

I have updated the draft PR with the usage of this "getGetter" function in PIDML code, and you can take a look at file "pidONNXModel.h" in this PR https://github.com/AliceO2Group/O2Physics/pull/7371

ktf replied:
I did not explain myself. I mean the actual `labelView` copy you create here. Why can't you use `columnLabel` directly?

mytkom replied:
Oh, I am sorry I misunderstood, my bad. You are right this copy is unnecessary.

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1988,63 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+
+  if constexpr (std::is_arithmetic_v<typename C::type> && std::is_convertible_v<typename C::type, R>) {
+    if constexpr (o2::soa::is_dynamic_v<C>) {
+      if constexpr (o2::framework::pack_size(typename C::bindings_t{}) == o2::framework::pack_size(typename C::callable_t::args{})) {
```
I would make the whole constexpr logic a C++20 constraint (i.e.`requires requires is_arithmetic<typename C::type>...` so that we get a compilation error when the getter cannot be retrieved. Do you actually have any use for the case where you return a nullptr?

mytkom replied:
I can try filter some column types out of this iteration and then always return getter (because I would know that it can be retrieved), it would be safer than returning nullptr. Thanks

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1988,63 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, typename C>
+R getColumnValue(const T& rowIterator)
+{
+  return static_cast<R>(static_cast<C>(rowIterator).get());
```
Do you really need the two static_casts?

mytkom replied:
Even in my use case, columns have different types, and this cast to R is needed to make a common "ColumnGetterFunction" type. The first static cast chooses a correct `get` function from `rowIterator`. Maybe I could skip this R cast explicitly, and it would be done automatically (but I need to test if it would work like that). Would it look tidier, then?

ktf replied:
I think my point is more about the C cast. I would assume that if you can do `static_cast<R>(static_cast<C>(x))`, you can also do `static_cast<R>(x)` directly, no?

mytkom replied:
R is short for common return type, which can be ex. float. This C cast is only needed to use the `get()` member method of the dynamic/persistent column, and then it is called; its return value is then cast to R type. So I can't cast it directly, because I do not know what I would cast

ktf replied:
Ok, understood the usecase. Shouldn't R be derivable from C?

mytkom replied:
As is my first comment, R is needed for a common return type, to be able to store pointer to function obtained in runtime. In compile time we do not know which column label (so also column::type) would match columnLabel from JSON. In ML use case we would use only floats, so this cast to common return type is desired.

ktf replied:
Ok, I understand. No further comments on this for now.

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1988,63 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+  if constexpr (std::is_arithmetic_v<typename C::type> && std::is_convertible_v<typename C::type, R>) {
+    if constexpr (o2::soa::is_dynamic_v<C>) {
+      if constexpr (o2::framework::pack_size(typename C::bindings_t{}) == o2::framework::pack_size(typename C::callable_t::args{})) {
+        if (std::string_view(&columnLabel[1]) == std::string_view(C::columnLabel()) || columnLabel == std::string_view(C::columnLabel())) {
```
Why do you need the stringview? Is columnLabel a char*? If yes, then you should probably use strcmp which will avoid computing the size upfront.

mytkom replied:
That's great suggestion, I will use strcmp

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1988,63 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+        }
+      }
+    } else if constexpr (o2::soa::is_persistent_v<C> && !o2::soa::is_index_column_v<C>) {
+      if (columnLabel == std::string_view(C::columnLabel())) {
```
Same as above.

ktf commented:
@mytkom thank you for your work. See inline comments. I might have more once those are resolved.

mytkom committed:
style improvements + using concepts for conditional code and selecting types

alibuild committed:
Please consider the following formatting changes

mytkom commented:
@ktf 
I have finally made the changes requested in your last review. Could you take a look at the current implementation?

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -2730,6 +2799,7 @@ DECLARE_SOA_ITERATOR_METADATA();
     }                                                                                                                      \
                                                                                                                            \
     using bindings_t = typename o2::framework::pack<Bindings...>;                                                          \
+    using bindings_types_t = typename o2::framework::pack<typename Bindings::type...>;                                     \
```
I think `bindings_types_t` is actually not used, is it?

mytkom replied:
Yes, I forgot to remove it at the end. Thanks

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1989,74 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, dynamic_with_common_getter<R> C>
+ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& columnLabel, bool& found)
+{
+  if (std::strcmp(&columnLabel[1], C::columnLabel()) == 0 || std::strcmp(columnLabel.data(), C::columnLabel()) == 0) {
```
since it's a string_view, you should probably use `strncmp`, no? You are not guaranteed it's null terminated.

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1989,74 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+  (void)((func = createGetterPtr<R, T, Cs>(columnLabel, found), found) || ...);
+
+  if (!found) {
+    throw std::runtime_error("Getter for provided columnLabel not found!");
```
Please use `o2::framework::runtime_error`. Also, what's is the advantage having `found` and not throwing directly?

mytkom replied:
How can I throw it directly? I need to iterate over all the columns, and only then I know that it is not found. I will use o2::framework::runtime_error

ktf replied:
Yes, you are right, however I think you can simplify this by returning nullptr when it does not match.

mytkom committed:
remove bindings_types_t

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1989,74 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, persistent_with_common_getter<R> C>
+ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& columnLabel, bool& found)
+{
+  if (std::strcmp(columnLabel.data(), C::columnLabel()) == 0) {
+    found = true;
+  }
+
+  return &getColumnValue<R, T, C>;
```
```suggestion
  return std::strcmp(columnLabel.data(), C::columnLabel()) ? nullptr : &getColumnValue<R, T, C>;
```

ktf replied:
and then it just becomes a matter of setting the func and checking if it's not 0, no?

mytkom committed:
use strncmp and use nullptr as fallback

alibuild committed:
Please consider the following formatting changes

mytkom commented:
@ktf 
I simplified this using `nullptr` as you advised; now I wonder if this `strncmp` in my implementation is not making it overcomplicated for such a comparison. Could you take a look and give me your opinion on this? Maybe using string + strcmp for `str.data()` instead of string_view would be more readable and performance drawback should not be so heavy.

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1990,77 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, persistent_with_common_getter<R> C>
+ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& columnLabel)
+{
+  const size_t n = columnLabel.size();
+
+  if (n == 0 || n != strlen(C::columnLabel())) {
+    return nullptr;
+  }
+
+  return (std::strncmp(columnLabel.data(), C::columnLabel(), n)) ? nullptr : &getColumnValue<R, T, C>;
```
Am I missing something or it could be actually simplified to:

```suggestion
  return std::strncmp(columnLabel.data(), C::columnLabel(), columnLabel.size()) ? nullptr : &getColumnValue<R, T, C>;
```

`C::columnLabel()` is guaranteed to be null terminated, no? The issue is the string_view.

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1990,77 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+  { t.get() } -> std::convertible_to<R>;
+};
+
+template <typename T, typename R>
+using with_common_getter_t = typename std::conditional<persistent_with_common_getter<T, R> || dynamic_with_common_getter<T, R>, std::true_type, std::false_type>::type;
```
This you can probably move inside where it's used.

mytkom replied:
I moved it just before function where it is used. What did you mean by inside? Can I declare a template inside a function?

mytkom replied:
@ktf

mytkom committed:
use strcmp

mytkom committed:
fix dynamic column logic

alibuild committed:
Please consider the following formatting changes

alibuild commented:
Error while checking build/O2/fullCI for 6e55f8da6a8592a4f7cf9af6cb10c5a5b86de16f at 2024-10-21 18:16:
```
## sw/BUILD/xjalienfs-latest/log
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.


## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-sim-challenge-test-latest/log
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/46}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/47}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/48}
./digi.log[ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/49}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/46}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/47}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/48}
./digi.log[9304:internal-dpl-ccdb-backend]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/49}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/46}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/47}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/48}
./digi.log[9316:TPCDigitizer_7]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/49}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/37}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/38}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/40}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/42}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/43}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/45}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/46}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/47}
./digi.log[9302:internal-dpl-clock]: [ERROR] Found duplicate input binding with different spec.:collisioncontext {SIM/COLLISIONCONTEXT/48}
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13498/6e55f8da6a8592a4f7cf9af6cb10c5a5b86de16f/build_O2_fullCI/pretty.html).

mytkom commented:
Good morning @ktf,
Could you review the changes from 2 weeks ago and answer my comment? I will be grateful!

mytkom commented:
@ktf
Would you have time to take a look at this PR once more?

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1990,77 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+    return nullptr;
+  }
+
+  return ((std::strcmp(&columnLabel[1], C::columnLabel()) && std::strcmp(columnLabel.data(), C::columnLabel()))) ? nullptr : &getColumnValue<R, T, C>;
```
sorry, but I still do not understand why it is so complicated and why you need to invoke strlen. C::columnLabel() is guaranteed to be null teminated and columnLabel has a size. Why not simply using `std::strncmp(&columnLabel[1], C::columnLabel(), columnLabel.size() - 1)`?

mytkom replied:
I simplified it a bit. What do you think?

mytkom replied:
I made some tests and spotted a bug; using strncmp like that will not work without a length check. Edge case:
columnLabel = "fP"
C::columnLabel = "fPx"
With strncmp with n = columnLabel.size() and if "fPx" occurs in an iteration before "fP", the wrong thing is returned.
Because of that, I decided to use the `==` operator between `string_view` and `const char*`. Performance should be similar to these complicated checks, and the code should have better readability. I hope this solution is sufficient!

mytkom committed:
refactor string comparison

mytkom committed:
fix edge case

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1990,75 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, dynamic_with_common_getter<R> C>
+ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& columnLabel)
+{
+  // allows user to use consistent formatting (with prefix) of all column labels
+  // by default there isn't 'f' prefix for dynamic column labels
+  bool isPrefixMatch = columnLabel.size() > 1 && columnLabel.substr(1) == C::columnLabel();
+  // check also exact match if user is aware of prefix missing
+  bool isExactMatch = columnLabel == C::columnLabel();
+
+  return (isPrefixMatch || isExactMatch) ? &getColumnValue<R, T, C> : nullptr;
```
Looks much better, thanks. As a final note, why doing the comparison twice? Is it optimised out? How about:

```C++
if (columnLabel.size() > 1) {
  return columnLabel.substr(1) == C::columnLabel() ? &getColumnValue<R, T, C> : nullptr;
}
// The only remaining case is columnLabel.size() <= 1
return columnLabel.size() == 1 && columnLabel == C::columnLabel() : &getColumnValue<R, T, C> : nullptr;
```

ktf commented on Framework/Core/include/Framework/ASoA.h:
```diff
@@ -1988,6 +1990,75 @@ std::tuple<typename Cs::type...> getRowData(arrow::Table* table, T rowIterator,
+template <typename R, typename T, dynamic_with_common_getter<R> C>
+ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& columnLabel)
+{
+  // allows user to use consistent formatting (with prefix) of all column labels
+  // by default there isn't 'f' prefix for dynamic column labels
+  bool isPrefixMatch = columnLabel.size() > 1 && columnLabel.substr(1) == C::columnLabel();
+  // check also exact match if user is aware of prefix missing
+  bool isExactMatch = columnLabel == C::columnLabel();
+
+  return (isPrefixMatch || isExactMatch) ? &getColumnValue<R, T, C> : nullptr;
```
Nevermind the previous post, which I accidentally deleted. You can still return immediately, no? Is it already optimised out? I do not think the compiler is allowed to do that.

```C++
if (columnLabel.size() > 1 && columnLabel.substr(1) == C::columnLabel()) {
  return &getColumnValue<R, T, C>;
}

if (columnLabel == C::columnLabel()) {
  return &getColumnValue<R, T, C>;
}

return nullptr;
```

mytkom replied:
You're right. It would be faster to return immediately. 

What do you think is not allowed by the compiler? Comparison of `const char*` and `string_view`?

mytkom replied:
@ktf can you review most previous implementation? What do you think?

mytkom committed:
use early return

mytkom committed:
formatting

mytkom commented:
The `string_view` API has an efficient `starts_with()` method, which would check exactly the case I want to handle and it is very readable. Then I think that creating `string_view` from `C::columnLabel()` is more performant, because `strlen()` is being executed once, only on creation. Without it, it is executed on every comparison, since in the worst case, there are 2 comparisons, it should increase worst-case performance. What do you think of this:
```cpp
template <typename R, typename T, persistent_with_common_getter<R> C>
ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& targetColumnLabel)
{
  return targetColumnLabel == C::columnLabel() ? &getColumnValue<R, T, C> : nullptr;
}

template <typename R, typename T, dynamic_with_common_getter<R> C>
ColumnGetterFunction<R, T> createGetterPtr(const std::string_view& targetColumnLabel)
{
  std::string_view columnLabel(C::columnLabel());

  // allows user to use consistent formatting (with prefix) of all column labels
  // by default there isn't 'f' prefix for dynamic column labels
  if (targetColumnLabel.starts_with("f") && targetColumnLabel.substr(1) == columnLabel) {
    return &getColumnValue<R, T, C>;
  }

  // check also exact match if user is aware of prefix missing
  if (targetColumnLabel == columnLabel) {
    return &getColumnValue<R, T, C>;
  }

  return nullptr;
}
```

mytkom committed:
use temporary string_view and starts_with method

alibuild commented:
Error while checking build/O2/fullCI for 962a8d2e9d6cd6cdffe78face3b4808664dc1e9e at 2024-12-10 06:29:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:38: error: 'is_dynamic_v' was not declared in this scope; did you mean 'is_dynamic_t'?
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2188:52: error: expected primary-expression before '>' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:38: error: label 'framework' referenced outside of any function
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:83: error: expected ';' before ')' token
/sw/SOURCES/O2/13498-slc8_x86-64/0/Framework/Core/include/Framework/ASoA.h:2190:139: error: expected unqualified-id before ')' token
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13498/962a8d2e9d6cd6cdffe78face3b4808664dc1e9e/build_O2_fullCI/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 962a8d2e9d6cd6cdffe78face3b4808664dc1e9e at 2024-12-04 12:06:
```
## sw/BUILD/O2-latest/log
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13498/962a8d2e9d6cd6cdffe78face3b4808664dc1e9e/build_O2_fullCI_slc9/pretty.html).

mytkom committed:
post rebase fixes

mytkom committed:
fix formatting

mytkom commented:
@ktf I have rebased PR, and now all checks are successful.

ktf commented:
Thank you for following this through. Merging this. Could you please move the helper functions which are only internal into an anonymous namespace?
